### PR TITLE
Add basic string output of the container

### PR DESCRIPTION
**Pull request by glibsm**
Bringing back what was removed as part of the refactor - an easy way to quickly see what's inside the container.

### CODE REVIEW

File: dig_test.go [hunk 1]
```diff
@@ -723,3 +723,22 @@ func TestInvokeFailures(t *testing.T) {
 		assert.Error(t, c.Invoke(func() (int, error) { return 42, errors.New("oh no") }))
 	})
 }
+
+func TestStringer(t *testing.T) {
+	c := New()
```
Comment (alsamylkin) [inline dig_test.go]: Add t.Parallel()?

File: stringer.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package dig
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// String representation of the entire Container
+func (c *Container) String() string {
```
Comment (alsamylkin) [inline stringer.go]: Why these functions are in a separate file? The different receiver type((pointer and value) in a String methods makes me to look in the other methods implementations.

File: dig_test.go [hunk 3]
```diff
@@ -723,3 +723,22 @@ func TestInvokeFailures(t *testing.T) {
...
+	b := &bytes.Buffer{}
+	fmt.Fprintln(b, c)
+	s := b.String()
+
+	assert.Contains(t, s, "*dig.A ->")
+	assert.Contains(t, s, "*dig.B ->")
+	assert.Contains(t, s, "func() (*dig.A, *dig.B)")
+
+	assert.Contains(t, s, "*dig.A =>")
+	assert.Contains(t, s, "*dig.B =>")
```
Comment (alsamylkin) [inline dig_test.go]: No test for `deps` and `constructor`?

File: stringer.go (hunk 2 above)
Comment (glibsm) [inline stringer.go]: This functionality is very self-contained even though it belongs to the same package.
I'd like to keep `dig.go` as tight as possible -- just the core functionality.

File: stringer_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	require.NoError(t, c.Provide(func() (*A, *B) { return &A{}, &B{} }))
+	require.NoError(t, c.Provide(func(*B) *C { return &C{} }))
+	require.NoError(t, c.Invoke(func(a *A) {}))
+
+	b := &bytes.Buffer{}
+	fmt.Fprintln(b, c)
+	s := b.String()
+
+	// all nodes are in the graph
+	assert.Contains(t, s, "*dig.A -> deps: []")
```
Comment (alsamylkin) [inline stringer_test.go]: nit: if you use assert many times, you can use: `assert := assert.New(t)`

File: stringer.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	b := &bytes.Buffer{}
+	fmt.Fprintln(b, "nodes: {")
+	for k, v := range c.nodes {
+		fmt.Fprintln(b, "\t", k, "->", v)
+	}
+	fmt.Fprintln(b, "}")
+
+	fmt.Fprintln(b, "cache: {")
+	for k, v := range c.cache {
+		fmt.Fprintln(b, "\t", k, "=>", v)
```
Comment (akshayjshah) [inline stringer.go]: Why the different arrows?

File: stringer.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	}
+	fmt.Fprintln(b, "}")
+
+	return b.String()
```
Comment (akshayjshah) [inline stringer.go]: This is *so* close to being JSON - why not implement `MarshalJSON` instead?

File: stringer.go (hunk 5 above)
Comment (glibsm) [inline stringer.go]: separating graph and cache outputs, open to suggestions

File: stringer.go (hunk 6 above)
Comment (glibsm) [inline stringer.go]: Yeah, but it's not the intention to have a JSON output :) The target audience is for people using dig, not someone parsing the output of the graph.

That said, having it be *actual* JSON might be beneficial if we ever wanted to actually try and parse it for some reason (dunno, codegen or something)

File: stringer.go (hunk 5 above)
Comment (akshayjshah) [inline stringer.go]: Works for me, just making sure it's intentional.

File: stringer.go (hunk 6 above)
Comment (akshayjshah) [inline stringer.go]: :+1: sure.

**Comment by glibsm** [comment]
Going to merge as-is since it's actually pretty useful even while writing new features of dig and unit testing.

We can iterate and convert to *actually* producing JSON later if we feel like it

**Pull request merged**